pkg/controller/direct/bigquerydatatransfer/bigquerydatatransferconfig_controller.go

```diff
@@ -0,0 +1,351 @@
+	"google.golang.org/api/option"
+	"google.golang.org/protobuf/types/known/fieldmaskpb"
+	"k8s.io/apimachinery/pkg/apis/meta/v1/unstructured"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	ctrlName      = "bigquerydatatransfer-controller"
+	serviceDomain = "//bigquerydatatransfer.googleapis.com"
+)
+
+func init() {
+	registry.RegisterModel(krm.GroupVersionKind, NewModel)
+}
+
+func NewModel(ctx context.Context, config *config.ControllerConfig) (directbase.Model, error) {
+	return &model{config: *config}, nil
+}
+
+var _ directbase.Model = &model{}
+
+type model struct {
+	config config.ControllerConfig
+}
+
+func (m *model) client(ctx context.Context) (*gcp.Client, error) {
+	var opts []option.ClientOption
+	if m.config.UserAgent != "" {
+		opts = append(opts, option.WithUserAgent(m.config.UserAgent))
+	}
+	if m.config.HTTPClient != nil {
+		opts = append(opts, option.WithHTTPClient(m.config.HTTPClient))
+	}
+	if m.config.UserProjectOverride && m.config.BillingProject != "" {
+		opts = append(opts, option.WithQuotaProject(m.config.BillingProject))
+	}
+
+	gcpClient, err := gcp.NewRESTClient(ctx, opts...)
+	if err != nil {
+		return nil, fmt.Errorf("building bigquerydatatransfer client: %w", err)
+	}
+	return gcpClient, err
+}
+
+func (m *model) AdapterForObject(ctx context.Context, reader client.Reader, u *unstructured.Unstructured) (directbase.Adapter, error) {
+	obj := &krm.BigQueryDataTransferConfig{}
+	if err := runtime.DefaultUnstructuredConverter.FromUnstructured(u.Object, &obj); err != nil {
+		return nil, fmt.Errorf("error converting to %T: %w", obj, err)
+	}
+
+	// Get ResourceID
+	resourceID := direct.ValueOf(obj.Spec.ResourceID)
+	if resourceID == "" {
+		serviceGeneratedID, err := parseServiceGeneratedIDFromExternalRef(obj)
+		if err != nil {
+			return nil, err
+		}
+		resourceID = serviceGeneratedID
+	}
+	// TODO(jingyih): need to figure out how to deal with resourceID. It is service generated.
+
+	location := obj.Spec.Location
+	if location == "" {
+		return nil, fmt.Errorf("cannot resolve location")
+	}
+
+	// Resolve PubSubTopic Ref
+	if obj.Spec.PubSubTopicRef != nil {
+		topic, err := refv1beta1.ResolvePubSubTopic(ctx, reader, obj, obj.Spec.PubSubTopicRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.PubSubTopicRef.External = topic.String()
+	}
+
+	// Resolve BigQueryDataSet Ref
+	if obj.Spec.DatasetRef != nil {
+		dataset, err := refv1beta1.ResolveBigQueryDataset(ctx, reader, obj, obj.Spec.DatasetRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.DatasetRef.External = dataset.String()
+	}
+
+	// Resolve KMSCryptoKey Ref
+	if obj.Spec.EncryptionConfiguration != nil {
+		kmsCryptoKeyRef, err := refv1beta1.ResolveKMSCryptoKeyRef(ctx, reader, obj, obj.Spec.EncryptionConfiguration.KmsKeyRef)
+		if err != nil {
+			return nil, err
+		}
+		obj.Spec.EncryptionConfiguration.KmsKeyRef = kmsCryptoKeyRef
+	}
```

Comment:
I think we shall move the non-parent depedencies in the create and update calls.

Reply:
Otherwise the presence of the dependency would block the target resource

Reply:
IIUC, it mainly affects `Find()` call? In other words, if the dependencies are not present, `Create()` and `Update()` will be blocked anyway, but `Find()` should not be blocked.

Reply:
It affects `Delete`. If a reference is missing, it cannot delete the object.

Reply:
I don't think this is a blocker issue.